backend/server/src/services/tournamentService.ts

```diff
@@ -67,4 +100,262 @@ export class TournamentService {
+          });
+        }
+        currentRoundPlayers.push(winnerId);
+      }
+    }
+
+    for (let i = 0; i < playerIds.length; i += 2) {
+      // only works in power of 2.
+      matches.push({
+        players: [
+          { id: playerIds[i], points: [], color: "red" },
+          { id: playerIds[i + 1], points: [], color: "white" }
+        ],
+        type: "playoff", // decide what this is going to be.
+        admin: null, // this has to be an ID.
+        elapsedTime: 0,
+        timerStartedTimestamp: null
+        // Set other Match fields as necessary
+      });
+    }
+    return matches;
+  }
+
+  // private utility function
+  private isPowerOfTwo(n: number): boolean {
+    if (n <= 0) {
+      return false;
+    }
+    return (n & (n - 1)) === 0;
+  }
+
+  // private utility function
+  private calculateRoundRobinMatches(playerCount: number): number {
+    if (playerCount < 2) {
+      throw new BadRequestError({
+        message:
+          "At least two players are required for a round robin tournament."
+      });
+    }
+    return (playerCount * (playerCount - 1)) / 2;
+  }
+
+  private async tournamentToObject(
+    tournamentObject: Tournament & Document
+  ): Promise<Tournament & Document> {
+    const playerPromises = tournamentObject.players.map(
+      async (playerId) =>
+        await UserModel.findById(playerId)
+          .select("firstName lastName _id")
+          .exec()
+    );
+
+    const players = await Promise.all(playerPromises);
+    tournamentObject.playerDetails = players.map((player) =>
+      player !== null && player !== undefined
+        ? {
+            firstName: player.firstName,
+            lastName: player.lastName,
+            id: player._id.toString()
+          }
+        : {
+            firstName: "",
+            lastName: "",
+            id: null
+          }
+    );
+
+    const matchPromises = tournamentObject.matchSchedule.map(
+      async (matchId) => await MatchModel.findById(matchId).exec()
+    );
+
+    const matches = await Promise.all(matchPromises);
+    const extendedMatches: ExtendedMatch[] = [];
+
+    for (const match of matches) {
+      if (match !== null && match !== undefined) {
+        const playerDetailsPromises = match.players.map(
+          async (player) =>
+            await UserModel.findById(player.id)
+              .select("firstName lastName _id")
+              .exec()
+        );
+
+        const playersDetails = await Promise.all(playerDetailsPromises);
+        const extendedPlayers = playersDetails.map((player) => ({
+          id:
+            player !== null && player !== undefined
+              ? player._id.toString()
+              : null,
+          firstName:
+            player !== null && player !== undefined ? player.firstName : null,
+          lastName:
+            player !== null && player !== undefined ? player.lastName : null
+        }));
```

Comment:
Having a hard time understanding what is going on here

Reply:
My understanding here is that we just map the data . This would be much easier if we just worked with nullable objects instead of objects that have all properties nullable and undefined.

Reply:
Resolved. Initially adding cause Lint was annoying me.